MC-1296 Deploy Reaper capable of interaction with encrypted mgmt-api

rzvoncek opened the pull request:
<!--  Thanks for sending a pull request!  Before submitting:

1. Read our CONTRIBUTING.md guide
2. Rebase your PR if it gets out of sync with main
-->

**What this PR does**:
This PR changes the reconciliation process of k8ssandra cluster and reaper objects so that it becomes possible for one a single reaper insance to interact with several k8ssandra clusters with their mgmt-api communication encrypted (each using a different certificate).

**Which issue(s) this PR fixes**:
Fixes 

**Checklist**
- [x] Changes manually tested
- [x] Automated Tests added/updated
- [ ] Documentation added/updated
- [x] CHANGELOG.md updated (not required for documentation PRs)
- [ ] CLA Signed:  [DataStax CLA](https://cla.datastax.com/)

Miles-Garnsey reviewed (changes requested):
@rzvoncek when I run `CreateControlPlaneReaperAndDataCenter` from vs code, I'm seeing the test fail - 

```
--- FAIL: TestOperator (926.60s)
    --- FAIL: TestOperator/CreateControlPlaneReaperAndDataCenter (926.60s)
        /Users/milesgarnsey/projects/k8ssandra-operator/test/e2e/e2e_framework.go:69: Using config file: /Users/milesgarnsey/projects/k8ssandra-operator/build/kind-kubeconfig
        /Users/milesgarnsey/projects/k8ssandra-operator/test/e2e/e2e_framework.go:77: Using control plane: kind-k8ssandra-0
        /Users/milesgarnsey/projects/k8ssandra-operator/test/e2e/e2e_framework.go:94: Using data planes: [kind-k8ssandra-0 kind-k8ssandra-1]
        /Users/milesgarnsey/projects/k8ssandra-operator/test/e2e/reaper_test.go:350: check that Reaper reaper1 in cluster kind-k8ssandra-0 is ready
        /Users/milesgarnsey/projects/k8ssandra-operator/test/e2e/suite_test.go:1805: check that datacenter enc-mgmt-dc1 in cluster kind-k8ssandra-0 is ready
        /Users/milesgarnsey/projects/k8ssandra-operator/test/e2e/suite_test.go:1805: check that datacenter enc-mgmt-2-dc1 in cluster kind-k8ssandra-0 is ready
        /Users/milesgarnsey/projects/k8ssandra-operator/test/e2e/reaper_test.go:321: Verify Reaper received k8ssandra-cluster secrets
        /Users/milesgarnsey/projects/k8ssandra-operator/test/e2e/suite_test.go:1951: 
            	Error Trace:	/Users/milesgarnsey/projects/k8ssandra-operator/test/e2e/suite_test.go:1951
            	            				/Users/milesgarnsey/projects/k8ssandra-operator/test/e2e/reaper_test.go:322
            	            				/Users/milesgarnsey/projects/k8ssandra-operator/test/e2e/suite_test.go:481
            	Error:      	Received unexpected error:
            	            	secrets "reaper1-truststores" not found
            	Test:       	TestOperator/CreateControlPlaneReaperAndDataCenter
            	Messages:   	failed to get reaper's truststore secret
        /Users/milesgarnsey/projects/k8ssandra-operator/test/e2e/suite_test.go:776: failed to delete namespace: context deadline exceeded
FAIL
FAIL	github.com/k8ssandra/k8ssandra-operator/test/e2e	927.122s
```

Is there something extra I need to do to make this work?

rzvoncek commented:
This led me to notice the test I added did not actually ran with the CI ... I've added it and it now ran [here](https://github.com/k8ssandra/k8ssandra-operator/actions/runs/11498842640/job/32005589775?pr=1421). It seems to have passed. The PR adds a bunch of stuff, but it seems to be all present in the changed files.

So I'm not sure why this would fail ... when running from an IDE. I always run the e2e tests from shell, with:

```
e2e_test="TestOperator/CreateControlPlaneReaperAndDataCenter"
make single-up
make E2E_TEST="$e2e_test" e2e-test
```

Could you please try that way?

Miles-Garnsey commented on controllers/k8ssandra/reaper.go:
```diff
@@ -267,6 +275,22 @@ func getSingleReaperDcName(kc *api.K8ssandraCluster) string {
 	return ""
 }
 
+func usesHttpAuth(kc *api.K8ssandraCluster, actualDc *cassdcapi.CassandraDatacenter) (bool, *cassdcapi.ManagementApiAuthManualConfig) {
```
Suggestion: this looks like a getter with some defaulting behaviour. Can I suggest that it should be a method on the `ManagementApiAuth` type.

Miles-Garnsey replied:
Understood, apologies I didn't realise it was in cass-operator.

Miles-Garnsey commented on controllers/k8ssandra/reaper.go:
```diff
@@ -290,3 +314,69 @@ func (r *K8ssandraClusterReconciler) addClusterToExternalReaper(
+	kc *api.K8ssandraCluster,
+	actualDc *cassdcapi.CassandraDatacenter,
+	logger logr.Logger,
+	authConfig *cassdcapi.ManagementApiAuthManualConfig,
```
Suggestion: could you get this from the kc object to avoid the additional parameter? This function has a lot already.

rzvoncek replied:
I'm sorry but I don't think removing the `authConfig` function parameter leads to simpler `addManagementApiSecretsToReaper` function. 
The reason is that the `authConfig` passed into this function comes from the `usesHttpAuth` function, which picks it from one of the two possible palces in the `K8ssandraCluster` spec. I'd have to

Miles-Garnsey replied:
OK, closing as this was only a suggestion

Miles-Garnsey commented on controllers/k8ssandra/reaper.go:
```diff
@@ -290,3 +314,69 @@ func (r *K8ssandraClusterReconciler) addClusterToExternalReaper(
+
+	reaperNamespace := kc.Spec.Reaper.ReaperRef.Namespace
+	if reaperNamespace == "" {
+		reaperNamespace = kc.Namespace
```
Question: Is this the right default? It seems to me like by default Reaper will deploy to the mission control namespace as a static resource. But I kind of prefer this logic because it mirrors what we do elsewhere more closely...

rzvoncek replied:
It is indeed very likely we'll be deploying Reaper in the `mission-control` namespace. But I also think we should not be hardcoing the defaults of that namespace anywhere. I think it once bit me already, althought I don't remember where.
So I went for waht we do elsewhere. We seem to be in agreement, so I'll go ahead and resolve this bit.

Miles-Garnsey commented on pkg/reaper/secrets.go:
```diff
@@ -92,3 +92,7 @@ func secretToEnvVars(secret *corev1.Secret, envUsernameParam, envPasswordParam s
 	return &usernameEnvVar, &passwordEnvVar, nil
 }
+
+func GetTruststoresSecretName(reaperName string) string {
```
Suggestion: consider abbreviating this to -ts (like we do -ks for keystores) because otherwise the name will be too long for the API server.

rzvoncek replied:
Pushed a commit doing this.

Miles-Garnsey commented on controllers/k8ssandra/reaper.go:
```diff
@@ -290,3 +314,69 @@ func (r *K8ssandraClusterReconciler) addClusterToExternalReaper(
+	}
+
+	cs := &corev1.Secret{}
+	csName := authConfig.ClientSecretName + "-ks"
```
Issue: It looks like maybe you're re-using the same client secret currently used by cass-operator. Do you think we should maybe issue new secrets for Reaper for the sake of security?

rzvoncek replied:
I agree with Micke - and not only becausae that means I don't have to change anything 😃 
Generating and managing another secret for Reaper <-> mgmt-api seems like a big headache. Taking this ad absurdum, would we then need each client of mgmt-api have its own secret - and mgmt-api accept all of them. That does not sound correct.

Miles-Garnsey replied:
The principle we have followed elsewhere is that each application gets its own set of encryption materials. So one set for Cassandra, one set for management API, one set for Reaper, etc. 

@rzvoncek suggest you check this with Chris, I'm not personally fussy either way.

And @burmanm - can you please elaborate on your perspectives here so we can understand your reasoning.

Miles-Garnsey replied:
> This is not the correct way, as mgmt-api supports a single TLS cert (combination of tls.crt/ca.crt/tls.key files, not Java keystores). Thus, if Reaper doesn't use the cass-operator one, then Reaper will not be able to connect to mgmt-api.

@burmanm I've found your reasoning in your additional comment further below. I think you've misunderstood what I'm suggesting, I am suggesting issuing a new Reaper cert using the same issuer used for the management API cert. Because they will both share the same issuer, the existing management API truststore will trust the new cert issued for Reaper.

Whether it uses a Java keystore or pem encoded stuff isn't relevant here.

rzvoncek replied:
You both have valid points and it's not immediately clear how to proceed.

To move us forward, I'd like to point out that this PR is dealing with _using_ a certificate. It's not concerned what other uses that certificates have, nor where it came from. Genereating certs is mission-control, not k8ssandra, operator's job anyway.

If we decide we want dedicated certs for Reaper, we will need to track this with another ticket.

Miles-Garnsey replied:
The issue here is that you're not allowing for two different secrets to be used, I think this code assumes that the `authConfig.ClientSecretName` (which is the same secret used by management API) will always also be used by Reaper.

Miles-Garnsey commented on controllers/k8ssandra/reaper.go:
```diff
@@ -290,3 +314,69 @@ func (r *K8ssandraClusterReconciler) addClusterToExternalReaper(
+		return err
+	}
+
+	clusterName := cassdcapi.CleanupForKubernetes(actualDc.Spec.ClusterName)
```
Issue: I'm not sure this is the right logic - in particular clusterName is not always populated. Is there a function you can re-use in cass operator to figure out the resource names?

rzvoncek replied:
The `actualDc.Spec.ClusterName` is a field of `CassandraDatacenterSepc` in cass-operator. It's declared like this:
```
	// +kubebuilder:validation:MinLength=2
	ClusterName string `json:"clusterName"`
```
This looks like the field must be a string of at least two characters. Otherwise, I think, the cass-operator's webhook will prevent creating the DC. In other words, if a `CassandraDatacenter` object shows up, this field must be at least 2 chars long. Please correct me if I'm missing something.

Anyhow, we could also take the K8ssandraCluster object name, or `kc.Spec.Cassandra.ClusterName`. I'd say all these should be set to the same thing, but I don't really know.

Could someone else please chime in?

Miles-Garnsey replied:
Re-reading this... You're only using it for setting names for the truststore/keystore files that get materialised in the secret right? So this will just affect the keys in the aggregated secret that contains all of the encryption materials for the different Cassandra clusters? 

It may not matter to much if that's right, although I'd still suggest you use something like the namespace-name of the k8ssandra cluster since otherwise multiple clusters with the same name in different namespaces can overwrite each other...

rzvoncek replied:
This is correct. I'll make a fix.

rzvoncek replied:
After some investigation, and a chat with Alex, we concluded that this is, in fact, a blocking issue.
We need a proper way of making Reaper (and its truststore) capable to tell two equally named clusters from different namespaces apart. It's not just for encryption, but for monitoring also.

We'll be putting this ticket on hold until we fix this in Reaper.

Miles-Garnsey commented on controllers/k8ssandra/reaper.go:
```diff
@@ -290,3 +314,69 @@ func (r *K8ssandraClusterReconciler) addClusterToExternalReaper(
+	_, hasTruststore := tss.Data[clustersTruststore]
+	_, hasKeystore := tss.Data[clustersKeystore]
+
+	if hasTruststore && hasKeystore {
```
Issue: Do you need to ensure that the content is correct? What if the secrets have rotated?

rzvoncek replied:
This is a good catch. The rotation is not currently handled. I've made [MC-1367](https://datastax.jira.com/browse/MC-1367) to add that. 
What other correctnes checks could you think of ?

Miles-Garnsey replied:
Fingerprints are probably sufficient until you need to think about rotation. We can think more about rotating the certs and CAs down the track (although CA related work is on hold at present, so I'm not sure when that will happen).

Miles-Garnsey commented on controllers/reaper/reaper_controller.go:
```diff
@@ -484,3 +492,31 @@ func (r *ReaperReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	}
+	if err := r.Client.Get(ctx, sKey, s); err != nil {
+		if errors.IsNotFound(err) {
+			logger.Info("Creating Reaper's truststore ConfigMap", "ConfigMap", sKey)
```
Issue: this should say secret, not configmap.

rzvoncek replied:
Pushed a commit doing this.

Miles-Garnsey commented on controllers/reaper/reaper_controller.go:
```diff
@@ -484,3 +492,31 @@ func (r *ReaperReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	if err := r.Client.Get(ctx, sKey, s); err != nil {
+		if errors.IsNotFound(err) {
+			logger.Info("Creating Reaper's truststore ConfigMap", "ConfigMap", sKey)
+			if err = controllerutil.SetControllerReference(actualReaper, s, r.Scheme); err != nil {
```
Issue: I think you need `controllerutil.SetOwnerReference` as well.

burmanm replied:
You do not, SetControllerReference sets OwnerReference. It just sets the "controller" part in the OwnerReference. 

https://github.com/kubernetes-sigs/controller-runtime/blob/main/pkg/controller/controllerutil/controllerutil.go#L85

Miles-Garnsey replied:
OK, all good then. We'd set both in some other cases in the codebase, maybe that was wrong.

Miles-Garnsey commented on controllers/reaper/reaper_controller_test.go:
```diff
@@ -570,6 +572,72 @@ func testCreateReaperWithLocalStorageType(t *testing.T, ctx context.Context, k8s
+	assert.Equal(t, 0, len(truststoresSecret.Data))
+
+	// In this configuration, we expect Reaper to also have a mount for the http auth secrets
+	assert.Len(t, sts.Spec.Template.Spec.Containers[0].VolumeMounts, 3)
```
Issue: why do we need a deepcopy here? Also, I think you might be better off writing something more abstract, as right now any changes to the ordering of the mounts etc. will cause this test to fail.

rzvoncek replied:
I've removed the `DeepCopy()`. No idea why it was there.

rzvoncek replied:
I've also refactored the access by index. Now it finds the container by name first.

Miles-Garnsey commented on controllers/reaper/reaper_controller_test.go:
```diff
@@ -570,6 +572,72 @@ func testCreateReaperWithLocalStorageType(t *testing.T, ctx context.Context, k8s
+
+	assert.Eventually(t, func() bool {
+		err = k8sClient.Get(ctx, secretKey, truststoresSecret)
+		// again, we'd expect errors.IsNotFound(err) ...
```
Issue: can you double check this comment in light of cleanups not happening in envtests. And the one above too...

burmanm replied:
I'm not sure what you meant here, but there's no garbage collection in envtests. So no reference deletions happen as there's no such controller running in envtest. So the comment on line 612 is correct.

Miles-Garnsey replied:
I wasn't talking about 612, I was talking abot 624.

rzvoncek replied:
I went back and double checked this.

In the test, I go ahead and delete a Reaper object. This object is pointed to from a stateful set and from a secret via owner reference. The two are owned by the Reaper object. 

When writing this test, I did not know what precisely the _no gargage collection in envtests_ means. The Reaper object got deleted, so I was expecting the other things to disapear too - because that's what they do in an actual cluster. However, without the garbage collection, they won't disappear and I have to delete them manually. 

As a step forward, I propse removing the two `assert.Eventually` checks because they are confusing.

rzvoncek replied:
I've removed the `assert.Eventually`s.

Miles-Garnsey commented on test/e2e/suite_test.go:
```diff
@@ -1933,24 +1934,47 @@ func createKeyspaceAndTable(
 	require.NoError(t, err, "failed to create table")
 }
 
+func verifyReaperSecrets(
```
Suggestion: it would be good to check the fingerprints on the certs to make sure they're what you expect.

Miles-Garnsey commented on test/testdata/fixtures/reaper-control-plane/reaper.yaml:
```diff
@@ -17,3 +17,5 @@ spec:
     enabled: false
   uiUserSecretRef:
     name: reaper-ui-secret
+  containerImage:
```
Question: Do we still need this or have we now done a release? I suggest you put a TODO comment here so that it gets (hopefully) picked up before release.

rzvoncek replied:
We only merged REAP-10 yesterday, we do not have a Reaper release with this just yet.

Miles-Garnsey reviewed (changes requested):
A few things:

1. I think you need to create new secrets for the sake of security, not re-use existing client secrets.
2. Unit/integration tests are currently failing.
3. I'm still not able to run `CreateControlPlaneReaperAndDataCenter` successfully. I also checked that other tests work. `AddDcToClusterSameDataplane` definitely works when run this way so it isn't just because we're running from the IDE.

If I'm understanding this correctly there are two sets of changes. One set in the Reaper controller creates mounts and an empty secret. Another set in the k8ssandra cluster controller adds the existing client encryption materials to Reaper's secret. 

I don't fully understand the API on the Reaper side yet, but this looks OK generally to me, just have a look at some of my comments throughout and the three main things I've highlighted above.

burmanm commented:
> I think you need to create new secrets for the sake of security, not re-use existing client secrets.

This is not the correct way, as mgmt-api supports a single TLS cert (combination of tls.crt/ca.crt/tls.key files, not Java keystores). Thus, if Reaper doesn't use the cass-operator one, then Reaper will not be able to connect to mgmt-api.
